File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/events/sdk_event.dart
import 'package:uuid/uuid.dart';

/// Event categories for routing and filtering
enum EventCategory {
  sdk,
  llm,
  stt,
  tts,
  vad,
  voice,
  model,
  device,
  network,
  storage,
  error,
  rag,
}

/// Event destination for routing
enum EventDestination {
  /// Send to both public EventBus and analytics
  all,

  /// Send only to public EventBus
  publicOnly,

  /// Send only to analytics (internal)
  analyticsOnly,
}

/// Base protocol for all SDK events.
///
/// Mirrors iOS `SDKEvent` protocol from RunAnywhere SDK.
/// Every event in the SDK should extend this class. The [destination] property
/// tells the router where to send the event:
/// - [EventDestination.all] (default) → EventBus + Analytics
/// - [EventDestination.publicOnly] → EventBus only
/// - [EventDestination.analyticsOnly] → Analytics only
///
/// Usage:
/// ```dart
/// EventPublisher.shared.track(LLMEvent.generationCompleted(...));
/// ```
abstract class SDKEvent {
  /// Unique identifier for this event instance
  String get id;

  /// Event type string (used for analytics categorization)
  String get type;

  /// Category for filtering/routing
  EventCategory get category;

  /// When the event occurred
  DateTime get timestamp;

  /// Optional session ID for grouping related events
  String? get sessionId => null;

  /// Where to route this event
  EventDestination get destination => EventDestination.all;

  /// Event properties as key-value pairs (for analytics serialization)
  Map<String, String> get properties => {};
}

/// Mixin providing default implementations for SDKEvent fields.
/// Similar to Swift protocol extensions.
mixin SDKEventDefaults implements SDKEvent {
  static const _uuid = Uuid();

  @override
  String get id => _uuid.v4();

  @override
  DateTime get timestamp => DateTime.now();

  @override
  String? get sessionId => null;

  @override
  EventDestination get destination => EventDestination.all;

  @override
  Map<String, String> get properties => {};
}

// ============================================================================
// SDK Initialization Events
// ============================================================================

/// SDK initialization events
abstract class SDKInitializationEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.sdk;
}

class SDKInitializationStarted extends SDKInitializationEvent {
  @override
  String get type => 'sdk.initialization.started';
}

class SDKInitializationCompleted extends SDKInitializationEvent {
  @override
  String get type => 'sdk.initialization.completed';
}

class SDKInitializationFailed extends SDKInitializationEvent {
  final Object error;

  SDKInitializationFailed(this.error);

  @override
  String get type => 'sdk.initialization.failed';

  @override
  Map<String, String> get properties => {'error': error.toString()};
}

// ============================================================================
// SDK Configuration Events
// ============================================================================

/// SDK configuration events
abstract class SDKConfigurationEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.sdk;
}

// ============================================================================
// SDK Generation Events (LLM)
// ============================================================================

/// SDK generation events
abstract class SDKGenerationEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.llm;

  static SDKGenerationStarted started({required String prompt}) {
    return SDKGenerationStarted(prompt: prompt);
  }

  static SDKGenerationCompleted completed({
    required String response,
    required int tokensUsed,
    required int latencyMs,
  }) {
    return SDKGenerationCompleted(
      response: response,
      tokensUsed: tokensUsed,
      latencyMs: latencyMs,
    );
  }

  static SDKGenerationFailed failed(Object error) {
    return SDKGenerationFailed(error);
  }

  static SDKGenerationCostCalculated costCalculated({
    required double amount,
    required double savedAmount,
  }) {
    return SDKGenerationCostCalculated(
      amount: amount,
      savedAmount: savedAmount,
    );
  }
}

class SDKGenerationStarted extends SDKGenerationEvent {
  final String prompt;

  SDKGenerationStarted({required this.prompt});

  @override
  String get type => 'llm.generation.started';

  @override
  Map<String, String> get properties => {'prompt_length': '${prompt.length}'};
}

class SDKGenerationCompleted extends SDKGenerationEvent {
  final String response;
  final int tokensUsed;
  final int latencyMs;

  SDKGenerationCompleted({
    required this.response,
    required this.tokensUsed,
    required this.latencyMs,
  });

  @override
  String get type => 'llm.generation.completed';

  @override
  Map<String, String> get properties => {
        'response_length': '${response.length}',
        'tokens_used': '$tokensUsed',
        'latency_ms': '$latencyMs',
      };
}

class SDKGenerationFailed extends SDKGenerationEvent {
  final Object error;

  SDKGenerationFailed(this.error);

  @override
  String get type => 'llm.generation.failed';

  @override
  Map<String, String> get properties => {'error': error.toString()};
}

class SDKGenerationCostCalculated extends SDKGenerationEvent {
  final double amount;
  final double savedAmount;

  SDKGenerationCostCalculated({
    required this.amount,
    required this.savedAmount,
  });

  @override
  String get type => 'llm.generation.cost_calculated';

  @override
  Map<String, String> get properties => {
        'amount': amount.toStringAsFixed(6),
        'saved_amount': savedAmount.toStringAsFixed(6),
      };
}

// ============================================================================
// SDK Model Events
// ============================================================================

/// SDK model events
abstract class SDKModelEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.model;

  static SDKModelLoadStarted loadStarted({required String modelId}) {
    return SDKModelLoadStarted(modelId: modelId);
  }

  static SDKModelLoadCompleted loadCompleted({required String modelId}) {
    return SDKModelLoadCompleted(modelId: modelId);
  }

  static SDKModelLoadFailed loadFailed({
    required String modelId,
    required Object error,
  }) {
    return SDKModelLoadFailed(modelId: modelId, error: error);
  }

  static SDKModelUnloadStarted unloadStarted({required String modelId}) {
    return SDKModelUnloadStarted(modelId: modelId);
  }

  static SDKModelUnloadCompleted unloadCompleted({required String modelId}) {
    return SDKModelUnloadCompleted(modelId: modelId);
  }

  static SDKModelDeleted deleted({required String modelId}) {
    return SDKModelDeleted(modelId: modelId);
  }

  // Download events
  static SDKModelDownloadStarted downloadStarted({required String modelId}) {
    return SDKModelDownloadStarted(modelId: modelId);
  }

  static SDKModelDownloadCompleted downloadCompleted(
      {required String modelId}) {
    return SDKModelDownloadCompleted(modelId: modelId);
  }

  static SDKModelDownloadFailed downloadFailed({
    required String modelId,
    required String error,
  }) {
    return SDKModelDownloadFailed(modelId: modelId, error: error);
  }

  static SDKModelDownloadProgress downloadProgress({
    required String modelId,
    required double progress,
  }) {
    return SDKModelDownloadProgress(modelId: modelId, progress: progress);
  }
}

class SDKModelLoadStarted extends SDKModelEvent {
  final String modelId;

  SDKModelLoadStarted({required this.modelId});

  @override
  String get type => 'model.load.started';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelLoadCompleted extends SDKModelEvent {
  final String modelId;

  SDKModelLoadCompleted({required this.modelId});

  @override
  String get type => 'model.load.completed';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelLoadFailed extends SDKModelEvent {
  final String modelId;
  final Object error;

  SDKModelLoadFailed({required this.modelId, required this.error});

  @override
  String get type => 'model.load.failed';

  @override
  Map<String, String> get properties => {
        'model_id': modelId,
        'error': error.toString(),
      };
}

class SDKModelUnloadStarted extends SDKModelEvent {
  final String modelId;

  SDKModelUnloadStarted({required this.modelId});

  @override
  String get type => 'model.unload.started';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelUnloadCompleted extends SDKModelEvent {
  final String modelId;

  SDKModelUnloadCompleted({required this.modelId});

  @override
  String get type => 'model.unload.completed';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelDeleted extends SDKModelEvent {
  final String modelId;

  SDKModelDeleted({required this.modelId});

  @override
  String get type => 'model.deleted';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelDownloadStarted extends SDKModelEvent {
  final String modelId;

  SDKModelDownloadStarted({required this.modelId});

  @override
  String get type => 'model.download.started';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelDownloadCompleted extends SDKModelEvent {
  final String modelId;

  SDKModelDownloadCompleted({required this.modelId});

  @override
  String get type => 'model.download.completed';

  @override
  Map<String, String> get properties => {'model_id': modelId};
}

class SDKModelDownloadFailed extends SDKModelEvent {
  final String modelId;
  final String error;

  SDKModelDownloadFailed({required this.modelId, required this.error});

  @override
  String get type => 'model.download.failed';

  @override
  Map<String, String> get properties => {
        'model_id': modelId,
        'error': error,
      };
}

class SDKModelDownloadProgress extends SDKModelEvent {
  final String modelId;
  final double progress;

  SDKModelDownloadProgress({required this.modelId, required this.progress});

  @override
  String get type => 'model.download.progress';

  @override
  Map<String, String> get properties => {
        'model_id': modelId,
        'progress': progress.toString(),
      };
}

// ============================================================================
// SDK Voice Events
// ============================================================================

/// SDK voice events
abstract class SDKVoiceEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.voice;

  static SDKVoiceListeningStarted listeningStarted() {
    return SDKVoiceListeningStarted();
  }

  static SDKVoiceListeningEnded listeningEnded() {
    return SDKVoiceListeningEnded();
  }

  static SDKVoiceSpeechDetected speechDetected() {
    return SDKVoiceSpeechDetected();
  }

  static SDKVoiceTranscriptionStarted transcriptionStarted() {
    return SDKVoiceTranscriptionStarted();
  }

  static SDKVoiceTranscriptionPartial transcriptionPartial(
      {required String text}) {
    return SDKVoiceTranscriptionPartial(text: text);
  }

  static SDKVoiceTranscriptionFinal transcriptionFinal({required String text}) {
    return SDKVoiceTranscriptionFinal(text: text);
  }

  static SDKVoiceResponseGenerated responseGenerated({required String text}) {
    return SDKVoiceResponseGenerated(text: text);
  }

  static SDKVoiceSynthesisStarted synthesisStarted() {
    return SDKVoiceSynthesisStarted();
  }

  static SDKVoiceAudioGenerated audioGenerated({required dynamic data}) {
    return SDKVoiceAudioGenerated(data: data);
  }

  static SDKVoiceSynthesisCompleted synthesisCompleted() {
    return SDKVoiceSynthesisCompleted();
  }

  static SDKVoicePipelineError pipelineError(Object error) {
    return SDKVoicePipelineError(error: error);
  }

  static SDKVoicePipelineStarted pipelineStarted() {
    return SDKVoicePipelineStarted();
  }

  static SDKVoicePipelineCompleted pipelineCompleted() {
    return SDKVoicePipelineCompleted();
  }
}

class SDKVoiceListeningStarted extends SDKVoiceEvent {
  @override
  String get type => 'voice.listening.started';
}

class SDKVoiceListeningEnded extends SDKVoiceEvent {
  @override
  String get type => 'voice.listening.ended';
}

class SDKVoiceSpeechDetected extends SDKVoiceEvent {
  @override
  String get type => 'voice.speech.detected';
}

class SDKVoiceTranscriptionStarted extends SDKVoiceEvent {
  @override
  String get type => 'voice.transcription.started';

  @override
  EventCategory get category => EventCategory.stt;
}

class SDKVoiceTranscriptionPartial extends SDKVoiceEvent {
  final String text;

  SDKVoiceTranscriptionPartial({required this.text});

  @override
  String get type => 'voice.transcription.partial';

  @override
  EventCategory get category => EventCategory.stt;

  @override
  Map<String, String> get properties => {'text': text};
}

class SDKVoiceTranscriptionFinal extends SDKVoiceEvent {
  final String text;

  SDKVoiceTranscriptionFinal({required this.text});

  @override
  String get type => 'voice.transcription.final';

  @override
  EventCategory get category => EventCategory.stt;

  @override
  Map<String, String> get properties => {'text': text};
}

class SDKVoiceResponseGenerated extends SDKVoiceEvent {
  final String text;

  SDKVoiceResponseGenerated({required this.text});

  @override
  String get type => 'voice.response.generated';

  @override
  Map<String, String> get properties => {'text_length': '${text.length}'};
}

class SDKVoiceSynthesisStarted extends SDKVoiceEvent {
  @override
  String get type => 'voice.synthesis.started';

  @override
  EventCategory get category => EventCategory.tts;
}

class SDKVoiceAudioGenerated extends SDKVoiceEvent {
  final dynamic data;

  SDKVoiceAudioGenerated({required this.data});

  @override
  String get type => 'voice.audio.generated';

  @override
  EventCategory get category => EventCategory.tts;
}

class SDKVoiceSynthesisCompleted extends SDKVoiceEvent {
  @override
  String get type => 'voice.synthesis.completed';

  @override
  EventCategory get category => EventCategory.tts;
}

class SDKVoicePipelineError extends SDKVoiceEvent {
  final Object error;

  SDKVoicePipelineError({required this.error});

  @override
  String get type => 'voice.pipeline.error';

  @override
  EventCategory get category => EventCategory.error;

  @override
  Map<String, String> get properties => {'error': error.toString()};
}

class SDKVoicePipelineStarted extends SDKVoiceEvent {
  @override
  String get type => 'voice.pipeline.started';
}

class SDKVoicePipelineCompleted extends SDKVoiceEvent {
  @override
  String get type => 'voice.pipeline.completed';
}

// ============================================================================
// SDK Device Events
// ============================================================================

/// SDK device events.
///
/// Mirrors iOS `DeviceEvent` from RunAnywhere SDK.
abstract class SDKDeviceEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.device;

  /// Factory method: device registered successfully
  static DeviceRegistered registered({required String deviceId}) {
    return DeviceRegistered(deviceId: deviceId);
  }

  /// Factory method: device registration failed
  static DeviceRegistrationFailed registrationFailed({required String error}) {
    return DeviceRegistrationFailed(error: error);
  }
}

class DeviceRegistered extends SDKDeviceEvent {
  final String deviceId;

  DeviceRegistered({required this.deviceId});

  @override
  String get type => 'device.registered';

  @override
  Map<String, String> get properties => {
        'device_id':
            deviceId.length > 8 ? '${deviceId.substring(0, 8)}...' : deviceId
      };
}

class DeviceRegistrationFailed extends SDKDeviceEvent {
  final String error;

  DeviceRegistrationFailed({required this.error});

  @override
  String get type => 'device.registration.failed';

  @override
  Map<String, String> get properties => {'error': error};
}

// ============================================================================
// SDK Storage Events
// ============================================================================

/// SDK storage events
abstract class SDKStorageEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.storage;

  /// Factory method: cache cleared
  static SDKStorageCacheCleared cacheCleared() {
    return SDKStorageCacheCleared();
  }

  /// Factory method: temp files cleaned
  static SDKStorageTempFilesCleaned tempFilesCleaned() {
    return SDKStorageTempFilesCleaned();
  }
}

class SDKStorageCacheCleared extends SDKStorageEvent {
  @override
  String get type => 'storage.cache.cleared';
}

class SDKStorageTempFilesCleaned extends SDKStorageEvent {
  @override
  String get type => 'storage.temp_files.cleaned';
}

// ============================================================================
// SDK RAG Events
// ============================================================================

/// SDK RAG (Retrieval-Augmented Generation) events.
///
/// Mirrors iOS `RAGEvents` from RunAnywhere SDK.
/// Published during the RAG pipeline lifecycle — creation, ingestion, and query.
abstract class SDKRAGEvent with SDKEventDefaults {
  @override
  EventCategory get category => EventCategory.rag;

  /// Pipeline created successfully.
  static SDKRAGPipelineCreated pipelineCreated() {
    return SDKRAGPipelineCreated();
  }

  /// Pipeline destroyed and resources released.
  static SDKRAGPipelineDestroyed pipelineDestroyed() {
    return SDKRAGPipelineDestroyed();
  }

  /// Document ingestion started.
  ///
  /// [documentLength] — character count of the document being ingested.
  static SDKRAGIngestionStarted ingestionStarted({
    required int documentLength,
  }) {
    return SDKRAGIngestionStarted(documentLength: documentLength);
  }

  /// Document ingestion completed.
  ///
  /// [chunkCount] — number of chunks created from the document.
  /// [durationMs] — time taken for ingestion in milliseconds.
  static SDKRAGIngestionComplete ingestionComplete({
    required int chunkCount,
    required double durationMs,
  }) {
    return SDKRAGIngestionComplete(
      chunkCount: chunkCount,
      durationMs: durationMs,
    );
  }

  /// RAG query started.
  ///
  /// [questionLength] — character count of the question (not the raw text).
  static SDKRAGQueryStarted queryStarted({required int questionLength}) {
    return SDKRAGQueryStarted(questionLength: questionLength);
  }

  /// RAG query completed with results.
  ///
  /// [answerLength] — character count of the generated answer.
  /// [chunksRetrieved] — number of chunks used as context.
  /// [retrievalTimeMs] — time taken for vector retrieval.
  /// [generationTimeMs] — time taken for LLM generation.
  /// [totalTimeMs] — total query time.
  static SDKRAGQueryComplete queryComplete({
    required int answerLength,
    required int chunksRetrieved,
    required double retrievalTimeMs,
    required double generationTimeMs,
    required double totalTimeMs,
  }) {
    return SDKRAGQueryComplete(
      answerLength: answerLength,
      chunksRetrieved: chunksRetrieved,
      retrievalTimeMs: retrievalTimeMs,
      generationTimeMs: generationTimeMs,
      totalTimeMs: totalTimeMs,
    );
  }

  /// RAG pipeline encountered an error.
  ///
  /// [message] — human-readable error description.
  static SDKRAGError error({required String message}) {
    return SDKRAGError(message: message);
  }
}

class SDKRAGPipelineCreated extends SDKRAGEvent {
  @override
  String get type => 'rag.pipeline.created';
}

class SDKRAGPipelineDestroyed extends SDKRAGEvent {
  @override
  String get type => 'rag.pipeline.destroyed';
}

class SDKRAGIngestionStarted extends SDKRAGEvent {
  final int documentLength;

  SDKRAGIngestionStarted({required this.documentLength});

  @override
  String get type => 'rag.ingestion.started';

  @override
  Map<String, String> get properties => {
        'document_length': '$documentLength',
      };
}

class SDKRAGIngestionComplete extends SDKRAGEvent {
  final int chunkCount;
  final double durationMs;

  SDKRAGIngestionComplete({
    required this.chunkCount,
    required this.durationMs,
  });

  @override
  String get type => 'rag.ingestion.complete';

  @override
  Map<String, String> get properties => {
        'chunk_count': '$chunkCount',
        'duration_ms': durationMs.toStringAsFixed(1),
      };
}

class SDKRAGQueryStarted extends SDKRAGEvent {
  final int questionLength;

  SDKRAGQueryStarted({required this.questionLength});

  @override
  String get type => 'rag.query.started';

  @override
  Map<String, String> get properties => {
        'question_length': '$questionLength',
      };
}

class SDKRAGQueryComplete extends SDKRAGEvent {
  final int answerLength;
  final int chunksRetrieved;
  final double retrievalTimeMs;
  final double generationTimeMs;
  final double totalTimeMs;

  SDKRAGQueryComplete({
    required this.answerLength,
    required this.chunksRetrieved,
    required this.retrievalTimeMs,
    required this.generationTimeMs,
    required this.totalTimeMs,
  });

  @override
  String get type => 'rag.query.complete';

  @override
  Map<String, String> get properties => {
        'answer_length': '$answerLength',
        'chunks_retrieved': '$chunksRetrieved',
        'retrieval_time_ms': retrievalTimeMs.toStringAsFixed(1),
        'generation_time_ms': generationTimeMs.toStringAsFixed(1),
        'total_time_ms': totalTimeMs.toStringAsFixed(1),
      };
}

class SDKRAGError extends SDKRAGEvent {
  final String message;

  SDKRAGError({required this.message});

  @override
  String get type => 'rag.error';

  @override
  EventDestination get destination => EventDestination.all;

  @override
  Map<String, String> get properties => {
        'error': message,
      };
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/events/SDKEvent.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Minimal event protocol for SDK events.
 * All event logic and definitions are in C++ (rac_analytics_events.h).
 * This Kotlin interface only provides the interface for bridged events.
 *
 * Mirrors Swift SDKEvent.swift exactly.
 */

package com.runanywhere.sdk.public.events

import com.runanywhere.sdk.public.extensions.RAG.RAGResult
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

// MARK: - Event Destination

/**
 * Where an event should be routed (mirrors C++ rac_event_destination_t).
 * Mirrors Swift EventDestination exactly.
 */
enum class EventDestination {
    /** Only to public EventBus (app developers) */
    PUBLIC_ONLY,

    /** Only to analytics/telemetry (backend) */
    ANALYTICS_ONLY,

    /** Both destinations (default) */
    ALL,
}

// MARK: - Event Category

/**
 * Event categories for filtering/grouping (mirrors C++ categories).
 * Mirrors Swift EventCategory exactly.
 */
enum class EventCategory(
    val value: String,
) {
    SDK("sdk"),
    MODEL("model"),
    LLM("llm"),
    STT("stt"),
    TTS("tts"),
    VOICE("voice"),
    STORAGE("storage"),
    DEVICE("device"),
    NETWORK("network"),
    ERROR("error"),
    RAG("rag"),
}

// MARK: - SDK Event Interface

/**
 * Minimal interface for SDK events.
 *
 * Events originate from C++ and are bridged to Kotlin via EventBridge.
 * App developers can subscribe to events via EventBus.
 *
 * Mirrors Swift SDKEvent protocol exactly.
 */
interface SDKEvent {
    /** Unique identifier for this event instance */
    val id: String

    /** Event type string (from C++ event types) */
    val type: String

    /** Category for filtering/routing */
    val category: EventCategory

    /** When the event occurred (epoch milliseconds) */
    val timestamp: Long

    /** Optional session ID for grouping related events */
    val sessionId: String?

    /** Where to route this event */
    val destination: EventDestination

    /** Event properties as key-value pairs */
    val properties: Map<String, String>
}

// MARK: - Base Event Implementation

/**
 * Base implementation of SDKEvent with default values.
 */
@OptIn(ExperimentalUuidApi::class)
data class BaseSDKEvent(
    override val type: String,
    override val category: EventCategory,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
    override val properties: Map<String, String> = emptyMap(),
) : SDKEvent

// MARK: - Specific Event Types

/**
 * SDK lifecycle events.
 */
@OptIn(ExperimentalUuidApi::class)
data class SDKLifecycleEvent(
    val lifecycleType: LifecycleType,
    val version: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
) : SDKEvent {
    override val type: String get() = "sdk.${lifecycleType.value}"
    override val category: EventCategory get() = EventCategory.SDK
    override val properties: Map<String, String>
        get() =
            buildMap {
                version?.let { put("version", it) }
            }

    enum class LifecycleType(
        val value: String,
    ) {
        INITIALIZED("initialized"),
        SHUTDOWN("shutdown"),
        ERROR("error"),
    }
}

/**
 * Model-related events.
 */
@OptIn(ExperimentalUuidApi::class)
data class ModelEvent(
    val eventType: ModelEventType,
    val modelId: String,
    val progress: Float? = null,
    val error: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
) : SDKEvent {
    override val type: String get() = "model.${eventType.value}"
    override val category: EventCategory get() = EventCategory.MODEL
    override val properties: Map<String, String>
        get() =
            buildMap {
                put("model_id", modelId)
                progress?.let { put("progress", it.toString()) }
                error?.let { put("error", it) }
            }

    enum class ModelEventType(
        val value: String,
    ) {
        DOWNLOAD_STARTED("download_started"),
        DOWNLOAD_PROGRESS("download_progress"),
        DOWNLOAD_COMPLETED("download_completed"),
        DOWNLOAD_FAILED("download_failed"),
        LOADED("loaded"),
        UNLOADED("unloaded"),
        DELETED("deleted"),
    }
}

/**
 * LLM-related events.
 */
@OptIn(ExperimentalUuidApi::class)
data class LLMEvent(
    val eventType: LLMEventType,
    val modelId: String? = null,
    val tokensGenerated: Int? = null,
    val latencyMs: Double? = null,
    val error: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
) : SDKEvent {
    override val type: String get() = "llm.${eventType.value}"
    override val category: EventCategory get() = EventCategory.LLM
    override val properties: Map<String, String>
        get() =
            buildMap {
                modelId?.let { put("model_id", it) }
                tokensGenerated?.let { put("tokens_generated", it.toString()) }
                latencyMs?.let { put("latency_ms", it.toString()) }
                error?.let { put("error", it) }
            }

    enum class LLMEventType(
        val value: String,
    ) {
        GENERATION_STARTED("generation_started"),
        GENERATION_COMPLETED("generation_completed"),
        GENERATION_FAILED("generation_failed"),
        STREAM_TOKEN("stream_token"),
        STREAM_COMPLETED("stream_completed"),
    }
}

/**
 * STT-related events.
 */
@OptIn(ExperimentalUuidApi::class)
data class STTEvent(
    val eventType: STTEventType,
    val modelId: String? = null,
    val transcript: String? = null,
    val confidence: Float? = null,
    val error: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
) : SDKEvent {
    override val type: String get() = "stt.${eventType.value}"
    override val category: EventCategory get() = EventCategory.STT
    override val properties: Map<String, String>
        get() =
            buildMap {
                modelId?.let { put("model_id", it) }
                transcript?.let { put("transcript", it) }
                confidence?.let { put("confidence", it.toString()) }
                error?.let { put("error", it) }
            }

    enum class STTEventType(
        val value: String,
    ) {
        TRANSCRIPTION_STARTED("transcription_started"),
        TRANSCRIPTION_COMPLETED("transcription_completed"),
        TRANSCRIPTION_FAILED("transcription_failed"),
        PARTIAL_RESULT("partial_result"),
    }
}

/**
 * TTS-related events.
 */
@OptIn(ExperimentalUuidApi::class)
data class TTSEvent(
    val eventType: TTSEventType,
    val voice: String? = null,
    val durationMs: Double? = null,
    val error: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
) : SDKEvent {
    override val type: String get() = "tts.${eventType.value}"
    override val category: EventCategory get() = EventCategory.TTS
    override val properties: Map<String, String>
        get() =
            buildMap {
                voice?.let { put("voice", it) }
                durationMs?.let { put("duration_ms", it.toString()) }
                error?.let { put("error", it) }
            }

    enum class TTSEventType(
        val value: String,
    ) {
        SYNTHESIS_STARTED("synthesis_started"),
        SYNTHESIS_COMPLETED("synthesis_completed"),
        SYNTHESIS_FAILED("synthesis_failed"),
        PLAYBACK_STARTED("playback_started"),
        PLAYBACK_COMPLETED("playback_completed"),
    }
}

/**
 * Error events.
 */
@OptIn(ExperimentalUuidApi::class)
data class ErrorEvent(
    val errorCode: String,
    val errorMessage: String,
    val errorCategory: String? = null,
    val component: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.ALL,
) : SDKEvent {
    override val type: String get() = "error.occurred"
    override val category: EventCategory get() = EventCategory.ERROR
    override val properties: Map<String, String>
        get() =
            buildMap {
                put("error_code", errorCode)
                put("error_message", errorMessage)
                errorCategory?.let { put("error_category", it) }
                component?.let { put("component", it) }
            }
}

/**
 * RAG-related events.
 * Mirrors Swift RAGEvent exactly.
 */
@OptIn(ExperimentalUuidApi::class)
data class RAGEvent(
    val eventType: RAGEventType,
    val documentLength: Int? = null,
    val chunkCount: Int? = null,
    val durationMs: Double? = null,
    val questionLength: Int? = null,
    val answerLength: Int? = null,
    val chunksRetrieved: Int? = null,
    val retrievalTimeMs: Double? = null,
    val generationTimeMs: Double? = null,
    val totalTimeMs: Double? = null,
    val errorMessage: String? = null,
    override val id: String = Uuid.random().toString(),
    override val timestamp: Long = System.currentTimeMillis(),
    override val sessionId: String? = null,
    override val destination: EventDestination = EventDestination.PUBLIC_ONLY,
) : SDKEvent {
    override val type: String get() = "rag.${eventType.value}"
    override val category: EventCategory get() = EventCategory.RAG
    override val properties: Map<String, String>
        get() = buildMap {
            documentLength?.let { put("documentLength", it.toString()) }
            chunkCount?.let { put("chunkCount", it.toString()) }
            durationMs?.let { put("durationMs", "%.1f".format(it)) }
            questionLength?.let { put("questionLength", it.toString()) }
            answerLength?.let { put("answerLength", it.toString()) }
            chunksRetrieved?.let { put("chunksRetrieved", it.toString()) }
            retrievalTimeMs?.let { put("retrievalTimeMs", "%.1f".format(it)) }
            generationTimeMs?.let { put("generationTimeMs", "%.1f".format(it)) }
            totalTimeMs?.let { put("totalTimeMs", "%.1f".format(it)) }
            errorMessage?.let { put("message", it) }
        }

    enum class RAGEventType(val value: String) {
        INGESTION_STARTED("ingestion.started"),
        INGESTION_COMPLETE("ingestion.complete"),
        QUERY_STARTED("query.started"),
        QUERY_COMPLETE("query.complete"),
        PIPELINE_CREATED("pipeline.created"),
        PIPELINE_DESTROYED("pipeline.destroyed"),
        ERROR("error"),
    }

    companion object {
        fun ingestionStarted(documentLength: Int) = RAGEvent(
            eventType = RAGEventType.INGESTION_STARTED,
            documentLength = documentLength,
        )

        fun ingestionComplete(chunkCount: Int, durationMs: Double) = RAGEvent(
            eventType = RAGEventType.INGESTION_COMPLETE,
            chunkCount = chunkCount,
            durationMs = durationMs,
        )

        fun queryStarted(question: String) = RAGEvent(
            eventType = RAGEventType.QUERY_STARTED,
            questionLength = question.length,
        )

        fun queryComplete(result: RAGResult) = RAGEvent(
            eventType = RAGEventType.QUERY_COMPLETE,
            answerLength = result.answer.length,
            chunksRetrieved = result.retrievedChunks.size,
            retrievalTimeMs = result.retrievalTimeMs,
            generationTimeMs = result.generationTimeMs,
            totalTimeMs = result.totalTimeMs,
        )

        fun pipelineCreated() = RAGEvent(eventType = RAGEventType.PIPELINE_CREATED)

        fun pipelineDestroyed() = RAGEvent(eventType = RAGEventType.PIPELINE_DESTROYED)

        fun error(message: String) = RAGEvent(
            eventType = RAGEventType.ERROR,
            errorMessage = message,
            destination = EventDestination.ALL,
        )
    }
}
